File: wp.nim
# nim c -o:wpnim -d:release --overflowChecks:off wp.nim

import algorithm, sequtils, strutils, tables

# CountTable.sort uses Shell sort and is very slow for millions of words
# so we use a normal table and sort it using algorithm.sort (merge sort)

var words = initTable[string, int]()

for line in stdin.lines:
  for word in line.splitWhitespace:
    if not words.hasKey(word):
      words[word]  = 1
    else:
      words[word] += 1

var ws = toSeq(words.pairs)
ws.sort(proc (x,y: tuple[word: string, cnt: int]): int =
  cmp(y[1], x[1]))

for word in ws.items:
  echo word[0], " ", word[1]

File: wp.lua
-- Compatible with Lua 5.1+
-- Somewhat balanced between readability and perfomance

local sort   = table.sort
local insert = table.insert
local gmatch = string.gmatch

local words = {}

for line in io.lines() do
  for word in gmatch(line, "%S+") do
    words[word] = (words[word] or 0) + 1
  end
end

local ws = {}
for word, count in pairs(words) do
  insert(ws, {word, count})
end

sort(ws, function(a, b)
  return a[2] > b[2]
end)

for i = 1, #ws do
  local entry = ws[i]
  print(entry[1] .. " " .. entry[2])
end
